runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/logging/Slf4jLogWriter.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution.logging;
+
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.LogEntry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link LogWriter} which uses an {@link Logger SLF4J Logger} as the underlying log backend.
+ *
+ * <p>Ignores the {@code timestamp}, {@code instruction reference}, {@code primitive transform
+ * reference}, and {@code thread} fields.
+ */
+public class Slf4jLogWriter implements LogWriter {
+  @Override
+  public void log(LogEntry entry) {
+    Logger log;
+    String location = entry.getLogLocation();
+    if (location != null) {
+      log = LoggerFactory.getLogger(location);
+    } else {
+      // TODO: Provide a useful default
+      log = LoggerFactory.getLogger("RemoteLog");
+    }
+    String message = entry.getMessage();
+    String trace = entry.getTrace();
+    switch (entry.getSeverity()) {
+      case ERROR:
+      case CRITICAL:
+        if (trace == null) {
+          log.error(message);
+        } else {
+          log.error("{} {}", message, trace);
+        }
+        break;
+      case WARN:
+        if (trace == null) {
+          log.warn(message);
+        } else {
+          log.warn("{} {}", message, trace);
+        }
+        break;
+      case INFO:
+      case NOTICE:
+        log.info(message);
+        break;
+      case DEBUG:
+        log.debug(message);
+        break;
+      case UNSPECIFIED:
+      default:
+        log.trace(message);
```

Comment:
I think default should map to INFO preceded by a warning message saying unknown log level, defaulting to INFO.

Reply:
The severity enum in `beam_fn_api.proto` says:

```
      // Trace level information, also the default log level unless
      // another severity is specified.
      TRACE = 1;
 ```

I can go change that easily, however

Reply:
The issue is that "default:" will translate everything to trace which isn't part of the switch statement to become trace.


Please drop the comment on beam_fn_api.proto about TRACE being the default if unspecified and add a comment to UNSPECIFIED saying that it will map onto TRACE. It used to be that TRACE was = 0 which meant that if the user didn't specify a log level explicitly then trace was the default. The comment was missed since all the enums moved to start at non zero values with UNSPECIFIED as the zero value.